src/combinatorics/simple_graph/handshake.lean

```diff
@@ -0,0 +1,262 @@
+  d₁.edge = d₂.edge ↔ d₁ = d₂ ∨ d₁ = d₂.rev :=
+begin
+  cases d₁ with s₁ t₁ h₁,
+  cases d₂ with s₂ t₂ h₂,
+  simp only [dart.edge, dart.rev_edge, dart.rev],
+  rw sym2.eq_iff,
+end
+
+variables (G)
+
+/-- For a given vertex `v`, the injective map from the incidence set at `v` to the darts there. --/
+def dart_from_neighbor_set (v : V) : G.neighbor_set v → G.dart :=
+λ w, ⟨v, w, w.property⟩
+
+lemma dart_from_neighbor_set_inj (v : V) : function.injective (G.dart_from_neighbor_set v) :=
+λ e₁ e₂ h, by { injection h with h₁ h₂, exact subtype.ext h₂ }
+
+section degree_sum
+variables [fintype V] [decidable_rel G.adj]
+
+lemma dart_fst_fiber [decidable_eq V] (v : V) :
+  univ.filter (λ d : G.dart, d.fst = v) = univ.image (G.dart_from_neighbor_set v) :=
+begin
+  ext d,
+  simp only [mem_image, true_and, mem_filter, set_coe.exists, mem_univ, exists_prop_of_true],
+  split,
+  { rintro rfl,
+    exact ⟨_, d.is_adj, dart.ext _ _ rfl rfl⟩, },
+  { rintro ⟨e, he, rfl⟩,
+    refl, },
+end
+
+lemma dart_fst_fiber_card_eq_degree [decidable_eq V] (v : V) :
+  (univ.filter (λ d : G.dart, d.fst = v)).card = G.degree v :=
+begin
+  have hh := card_image_of_injective univ (G.dart_from_neighbor_set_inj v),
+  rw [finset.card_univ, card_neighbor_set_eq_degree] at hh,
+  rwa dart_fst_fiber,
+end
+
+lemma dart_card_eq_sum_degrees : fintype.card G.dart = ∑ v, G.degree v :=
+begin
+  haveI h : decidable_eq V := by { classical, apply_instance },
+  simp only [←card_univ, ←dart_fst_fiber_card_eq_degree],
+  exact card_eq_sum_card_fiberwise (by simp),
+end
+
+variables [decidable_eq V]
+
+lemma dart_edge_fiber (d : G.dart) :
+  (univ.filter (λ (d' : G.dart), d'.edge = d.edge)) = {d, d.rev} :=
+finset.ext (λ d', by simpa using dart_edge_eq_iff d' d)
+
+lemma dart_edge_fiber_card (e : sym2 V) (h : e ∈ G.edge_set) :
+  (univ.filter (λ (d : G.dart), d.edge = e)).card = 2 :=
+begin
+  refine quotient.ind (λ p h, _) e h,
+  cases p with v w,
+  let d : G.dart := ⟨v, w, h⟩,
+  convert_to _ = finset.card {d, d.rev},
+  { rw [card_insert_of_not_mem, card_singleton],
+    rw [mem_singleton],
+    exact d.rev_ne.symm, },
+  congr,
+  apply G.dart_edge_fiber d,
+end
+
+lemma dart_card_eq_twice_card_edges : fintype.card G.dart = 2 * G.edge_finset.card :=
+begin
+  rw ←card_univ,
+  rw @card_eq_sum_card_fiberwise _ _ _ dart.edge _ G.edge_finset
+    (λ d h, by { rw mem_edge_finset, apply dart.edge_mem }),
+  rw [←mul_comm, sum_const_nat],
+  intros e h,
+  apply G.dart_edge_fiber_card e,
+  rwa ←mem_edge_finset,
+end
+
+/-- The degree-sum formula.  This is also known as the handshaking lemma, which might
+more specifically refer to `simple_graph.card_odd_degree_vertices_is_even`. -/
+theorem sum_degrees_eq_twice_card_edges : ∑ v, G.degree v = 2 * G.edge_finset.card :=
+G.dart_card_eq_sum_degrees.symm.trans G.dart_card_eq_twice_card_edges
+
+end degree_sum
+
+
+section TODO_move
+
+lemma zmod_eq_zero_iff_even (n : ℕ) : (n : zmod 2) = 0 ↔ even n :=
```

Comment:
I think we usually make arguments implicit if they are on both sides of an `iff`.
```suggestion
lemma zmod_eq_zero_iff_even {n : ℕ} : (n : zmod 2) = 0 ↔ even n :=
```